File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/trabalhos_praticos/TP_03/questao_05/Show.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

char toLower(char c){
    char resp = c;

    if(c >= 'A' && c <= 'Z'){
        resp = c + 32;
    }

    return resp;
}

int Strleng(char str[]){
    int length = 0;
    while(str[length] != '\0'){
        length++;
    }
    return length;
}

void Strcopy(char dest[], char src[]){
    int i = 0;
    while(src[i] != '\0'){
        dest[i] = src[i];
        i++;
    }
    dest[i] = '\0';
}

int isEqual(char str1[], char str2[]){
    int i = 0;
    int resp = 1;
    int len1 = Strleng(str1);
    int len2 = Strleng(str2);

    if(len1 == len2){
        while(i < len1 && resp){
            if(str1[i] != str2[i]){
                resp = 0;
            }
            i++;
        }
    } else{
        resp = 0;
    }

    return resp;
}

typedef struct{
    int ano;
    char mes[20];
    int dia;
} Data;

void Init_Date(Data *data){
    data->ano = 0;
    Strcopy(data->mes, "NaN");
    data->dia = 0;
}

void parseDate(Data *data, const char *source){
    if(source == NULL || source[0]== '\0'){
        Strcopy(data->mes, "March");
        data->dia = 1;
        data->ano = 1900;
    } else{
        char aux[100] = "";
        int auxIndex = 0;
        int step = 0;
        int i = 0;
    
        while(source[i] != '\0'){
            if(source[i] != ' ' && source[i] != ','){
                aux[auxIndex++] = source[i];
                aux[auxIndex] = '\0';
            } else{
                if(step == 0){
                    Strcopy(data->mes, aux);
                } else if(step == 1){
                    data->dia = atoi(aux);
                }
                auxIndex = 0;
                aux[0] = '\0';
                step++;
            }
            i++;
        }
    
        if(!isEqual(aux, "")){
            data->ano = atoi(aux);
        }
    }
    
}

typedef struct{
    char show_id[20];
    char type[20];
    char title[200];
    char director[100];
    char **cast;
    int cast_count;     
    char country[100];
    Data data_added;
    int release_year;
    char rating[20];
    char duration[20];
    char **listed_in;
    int listed_count;
} Show;


void Show_init(Show *s){
    Strcopy(s->show_id, "NaN");
    Strcopy(s->type, "NaN");
    Strcopy(s->title, "NaN");
    Strcopy(s->director, "NaN");
    
    s->cast = (char **)malloc(sizeof(char *));
    s->cast[0] = (char *)malloc(20 * sizeof(char));
    Strcopy(s->cast[0], "NaN");
    s->cast_count = 1;
    
    Strcopy(s->country, "NaN");
    
    Init_Date(&s->data_added);
    
    s->release_year = 0;
    Strcopy(s->rating, "NaN");
    Strcopy(s->duration, "NaN");
    
    
    s->listed_in = (char **)malloc(sizeof(char *));
    s->listed_in[0] = (char *)malloc(20 * sizeof(char));
    Strcopy(s->listed_in[0], "NaN");
    s->listed_count = 1;
}

int compareStrings(char str1[], char str2[]){
    int i = 0;
    int result = 0;

    while(str1[i] != '\0' && str2[i] != '\0' && result == 0){
        char c1 = toLower(str1[i]);
        char c2 = toLower(str2[i]);
        if(c1 != c2){
            result = c1 -  c2;
        }
        i++;
    }

    if(result == 0){
        result = Strleng(str1) - Strleng(str2);
    }
    return result;
}

void OrdenarCast(char **cast, int size){
    int i, j;
    char temp[100];

    for(i = 0; i < size; i++){
        for(j = 0; j < size - 1; j++){
            if(compareStrings(cast[j], cast[j + 1]) > 0){
                Strcopy(temp, cast[j]);
                Strcopy(cast[j], cast[j + 1]);
                Strcopy(cast[j + 1], temp);
            }
        }
    }
}

void Show_Imprimir(Show *s){
    printf("=> %s ## %s ## %s ## %s ## [",
            s->show_id, s->title, s->type, s->director);

    for(int i = 0; i < s->cast_count; i++){
        printf("%s", s->cast[i]);
        if(i < s->cast_count - 1){
            printf(", ");
        }
    }

    printf("] ## %s ## %s %d, %d ## %d ## %s ## %s ## [",
           s->country, s->data_added.mes, s->data_added.dia,
           s->data_added.ano, s->release_year, s->rating, s->duration);

    for(int i = 0; i < s->listed_count; i++){
        printf("%s", s->listed_in[i]);
        if(i < s->listed_count - 1){
            printf(", ");
        }
    }

    printf("] ##\n");
}

void Show_Ler(Show *show, char* line){
    char temp[11][1000];
    int TMPindex[11] = {0};
    int i = 0;
    int campo = 0;

    while(line[i] != '\0' && campo < 11){
        if(line[i] == '"'){
            i++;
            while(line[i] != '"' || (line[i] == '"' && line[i + 1] == '"')){
                if(line[i] == '"' && line[i + 1] == '"'){
                    i += 2;
                } else{
                    temp[campo][TMPindex[campo]++] = line[i++];
                }
            }
            i++;
            if (line[i] == ',')
                i++;
            temp[campo][TMPindex[campo]] = '\0';
            campo++;
        } else{
            while(line[i] != ',' && line[i] != '\0'){
                temp[campo][TMPindex[campo]++] = line[i++];
            }
            if(line[i] == ','){
                i++;
            }
            temp[campo][TMPindex[campo]] = '\0';
            campo++;
        }
    }

    Strcopy(show->show_id, temp[0][0] ? temp[0] : "NaN");
    Strcopy(show->type, temp[1][0] ? temp[1] : "NaN");
    Strcopy(show->title, temp[2][0] ? temp[2] : "NaN");
    Strcopy(show->director, temp[3][0] ? temp[3] : "NaN");
    Strcopy(show->country, temp[5][0] ? temp[5] : "NaN");
    parseDate(&show->data_added, temp[6]);
    show->release_year = atoi(temp[7]);
    Strcopy(show->rating, temp[8][0] ? temp[8] : "NaN");
    Strcopy(show->duration, temp[9][0] ? temp[9] : "NaN");

    if(temp[4][0]){
        show->cast_count = 1;
        for(i = 0; temp[4][i]; i++){
            if(temp[4][i] == ','){
                show->cast_count++;
            }
        }

        show->cast = (char **)malloc(show->cast_count * sizeof(char *));
        for (i = 0; i < show->cast_count; i++)
        {
            show->cast[i] = (char *)malloc(100 * sizeof(char));
        }

        int IndexC = 0;
        int currentCast = 0;
        i = 0;
        while(temp[4][i]){
            if(temp[4][i] == ','){
                show->cast[currentCast][IndexC] = '\0';
                currentCast++;
                IndexC = 0;
                i++;
                while(temp[4][i] == ' '){
                    i++;
                }
            } else{
                show->cast[currentCast][IndexC++] = temp[4][i++];
            }
        }
        show->cast[currentCast][IndexC] = '\0';
        OrdenarCast(show->cast, show->cast_count);
    } else{
        show->cast_count = 1;
        show->cast = (char **)malloc(sizeof(char *));
        show->cast[0] = (char *)malloc(20 * sizeof(char));
        Strcopy(show->cast[0], "NaN");
    }

    if(temp[10][0]){
        show->listed_count = 1;
        for(i = 0; temp[10][i]; i++){
            if(temp[10][i] == ','){
                show->listed_count++;
            }
        }

        show->listed_in = (char **)malloc(show->listed_count * sizeof(char *));
        for(i = 0; i < show->listed_count; i++){
            show->listed_in[i] = (char *)malloc(100 * sizeof(char));
        }

        int IndexL = 0;
        int currentList = 0;
        i = 0;
        while (temp[10][i]){
            if(temp[10][i] == ','){
                show->listed_in[currentList][IndexL] = '\0';
                currentList++;
                IndexL = 0;
                i++;
                while(temp[10][i] == ' '){
                    i++;
                }
            } else{
                show->listed_in[currentList][IndexL++] = temp[10][i++];
            }
        }
        show->listed_in[currentList][IndexL] = '\0';
    } else{
        show->listed_count = 1;
        show->listed_in = (char **)malloc(sizeof(char *));
        show->listed_in[0] = (char *)malloc(20 * sizeof(char));
        Strcopy(show->listed_in[0], "NaN");
    }
}

void copyCast(Show *copy, Show *source){
    copy->cast_count = source->cast_count;
    copy->cast = (char**)malloc(copy->cast_count * sizeof(char*));
    
    for(int i = 0; i < copy->cast_count; i++){
        copy->cast[i] = (char*)malloc(100 * sizeof(char));
        Strcopy(copy->cast[i], source->cast[i]);
    }
}

void copyListed(Show *copy, Show *source){
    copy->listed_count = source->listed_count;
    copy->listed_in = (char**)malloc(copy->listed_count * sizeof(char*));
    
    for(int i = 0; i < copy->listed_count; i++){
        copy->listed_in[i] = (char*)malloc(100 * sizeof(char));
        Strcopy(copy->listed_in[i], source->listed_in[i]);
    }
}

void clone(Show *array, Show *source){
    Strcopy(array->show_id, source->show_id);
    Strcopy(array->type, source->type);
    Strcopy(array->title, source->title);
    Strcopy(array->director, source->director);
    Strcopy(array->country, source->country);
    array->release_year = source->release_year;
    Strcopy(array->rating, source->rating);
    Strcopy(array->duration, source->duration);

    array->data_added.ano = source->data_added.ano;
    array->data_added.dia = source->data_added.dia;
    Strcopy(array->data_added.mes, source->data_added.mes);

    copyCast(array,source);
    copyListed(array,source);
}

int Pesquisar_ID(Show shows[], int size, char* id){
    int resp = 0;
    int i = 0;

    while(i < size){
        if(isEqual(shows[i].show_id, id)){
            resp = i;
            i = size;
        }
        i++;
    }
    return resp;
}

int IsFim(char line[]){
    int resp = 1;

    if(line[0] == 'F' && line[1] == 'I' && line[2] == 'M'){
        resp = 0;
    }

    return resp;
}

void freeShow(Show *show){
    for(int i = 0; i < show->cast_count; i++){
        free(show->cast[i]);
    }
    free(show->cast);

    for(int i = 0; i < show->listed_count; i++){
        free(show->listed_in[i]);
    }
    free(show->listed_in);
}

int buscar(Show shows[], int size, char *id){
    int resp = 0;
    int i = 0;

    while(i < size){
        if(strcmp(shows[i].show_id, id) == 0){
            resp = i;
            i = size;
        }
        i++;
    }
    return resp;
}

typedef struct Celula{
	Show elemento;        
	struct Celula* prox; 
} Celula;

Celula* novaCelula(Show elemento){
   Celula* nova = (Celula*) malloc(sizeof(Celula));
   nova->elemento = elemento;
   nova->prox = NULL;
   return nova;
}

Celula* primeiro;
Celula* ultimo;

void start (){
    Show cabeca;
    primeiro = novaCelula(cabeca);
    ultimo = primeiro;
}

int tamanho(){
   int tamanho = 0;
   Celula* i;
   for(i = primeiro; i != ultimo; i = i->prox, tamanho++);
   return tamanho;
}

void inserirInicio(Show x){
    Celula* tmp = novaCelula(x);

    tmp->prox = primeiro->prox;
    primeiro->prox = tmp;
    if(primeiro == ultimo){                   
      ultimo = tmp;
    }
    tmp = NULL;
}

void inserirFim(Show x){
    ultimo->prox = novaCelula(x);
    ultimo = ultimo->prox;
}

void inserir(Show x, int pos){
   int tam = tamanho();

    if(pos < 0 || pos > tam){
        printf("Erro ao inserir posicao (%d/ tamanho = %d) invalida!", pos, tam);
        exit(1);
    } else if(pos == 0){
        inserirInicio(x);
    } else if(pos == tam){
        inserirFim(x);
    } else{
        int j;
        Celula* i = primeiro;
        for(j = 0; j < pos; j++, i = i->prox);

        Celula* tmp = novaCelula(x);
        tmp->prox = i->prox;
        i->prox = tmp;
        tmp = i = NULL;
   }
}

Show removerInicio(){
    if(primeiro == ultimo){
        printf("Erro ao remover!");
        exit(1);
    }

    Celula* tmp = primeiro;
    primeiro = primeiro->prox;
    Show resp = primeiro->elemento;
    tmp->prox = NULL;
    free(tmp);
    tmp = NULL;

    return resp;
}

Show removerFim(){
    if(primeiro == ultimo){
        printf("Erro ao remover!");
        exit(1);
    } 

    Celula* i;
    for(i = primeiro; i->prox != ultimo; i = i->prox);

    Show resp = ultimo->elemento;
    ultimo = i;
    free(ultimo->prox);
    i = ultimo->prox = NULL;

    return resp;
}

Show remover(int pos){
    Show resp;
    int tam = tamanho();

    if(primeiro == ultimo){
        printf("Erro ao remover (vazia)!");
        exit(1);
    } else if(pos < 0 || pos >= tam){
        printf("Erro ao remover posicao (%d/ tamanho = %d) invalida!", pos, tam);
        exit(1);
    } else if(pos == 0){
        resp = removerInicio();
    } else if(pos == tam - 1){
        resp = removerFim();
    } else{
        Celula* i = primeiro;
        int j;
        for(j = 0; j < pos; j++, i = i->prox);

        Celula* tmp = i->prox;
        resp = tmp->elemento;
        i->prox = tmp->prox;
        tmp->prox = NULL;
        free(tmp);
        i = tmp = NULL;
    }

    return resp;
}

void mostrar(){
    Celula* i;

    for(i = primeiro->prox; i != NULL; i = i->prox){
        Show_Imprimir(&i->elemento);
    }
}

Show procurar(Show shows[], int size, char *id){
    Show resp;
    int i = 0;

    while(i < size){
        if(strcmp(shows[i].show_id, id) == 0){
            resp = shows[i];
            i = size;
        }
        i++;
    }

    return resp;
}

void metodos(int j, Show shows[], int show_count){
    char comando [3];
    char id [5];
    int pos;

    while (j > 0) {
        scanf(" %s", comando);
        if(strcmp(comando,"II") == 0){
            scanf(" %s", id); 
            inserirInicio(procurar(shows,show_count, id));
        }else if(strcmp(comando,"IF") == 0){
            scanf(" %s", id);
            inserirFim(procurar(shows,show_count, id));
        }else if(strcmp(comando,"I*") == 0){
            scanf(" %d", &pos); 
            scanf(" %s", id);
            inserir(procurar(shows,show_count, id),pos);
        }else if(strcmp(comando,"RI") == 0) {
            printf("(R) %s\n", removerInicio().title);
        }else if(strcmp(comando,"RF") == 0) {
            printf("(R) %s\n", removerFim().title);
        }else{
            scanf(" %d", &pos);
            printf("(R) %s\n", remover(pos).title);
        }
        j--;
    }
}

int main(){
    Show shows[1368];
    char line[5000];
    int show_count = 0, x = 0;

    FILE *file = fopen("/tmp/disneyplus.csv", "r");
    if(!file){
        printf("Arquivo nao encontrado!");
        return 1;
    }

    fgets(line, sizeof(line), file);

    while(fgets(line, sizeof(line), file) && show_count < 1368){
        Show_init(&shows[show_count]);
        Show_Ler(&shows[show_count], line);
        show_count++;
    }
    fclose(file);

    char input[20];
    scanf("%s", input);
    start();

    while (!strcmp(input, "FIM") == 0){
        int index = buscar(shows, show_count, input);
        inserirFim(shows[index]);
        scanf(" %s", input);
    }

    int qtd;
    scanf("%d", &qtd);
    metodos(qtd, shows, show_count);
    mostrar();

    for(int i = 0; i < show_count; i++){
        freeShow(&shows[i]);
    }

    return 0;
}

File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/trabalhos_praticos/TP_04/questao_07/Show.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.*;

class Celula{
	public Show elemento;
    public Celula prox;

    Celula(Show elemento){
        this.elemento = elemento;
        this.prox = null;
    }

	Celula(Show elemento, Celula prox){
		this.elemento = elemento;
		this.prox = prox;
	}
}

class Lista{
	private Celula primeiro; 
	private Celula ultimo; 

	public Lista(){
        primeiro = new Celula(null);
        ultimo = primeiro;
    }

	public void mostrar(){
		System.out.print("[ ");
		for(Celula i = primeiro.prox; i != null; i = i.prox){
			System.out.print(i.elemento + " ");
		}
		System.out.println("] ");
	}

	public boolean pesquisar(Show elemento, Log log){
        boolean achou = false;
        for(Celula i = primeiro.prox; i != null && !achou; i = i.prox){
            log.incrementaCount();
            if(i.elemento.getTitle().equals(elemento.getTitle())){
                achou = true;
            }
        }
        return achou;
    }


	public void inserirInicio(Show elemento){
		Celula tmp = new Celula(elemento);
        tmp.prox = primeiro.prox;
		primeiro.prox = tmp;
		if(primeiro == ultimo){
			ultimo = tmp;
		}
        tmp = null;
	}

	public void inserirFim(Show elemento){
		Celula tmp = new Celula(elemento);
		ultimo.prox = tmp;
		ultimo = ultimo.prox;
        tmp = null;
	}

    public void inserirMeio(Show x, int posicao) throws Exception{
        Celula i;
        int cont;

        for(i = primeiro, cont = 0; (i.prox != ultimo && cont < posicao); i = i.prox, cont++);
		
		if(posicao < 0 || posicao > cont + 1){
			throw new Exception("Erro ao inserir (posicao " + posicao + "(cont = " + cont + ") invalida)!");
        } else if(posicao == cont + 1){
            inserirFim(x);
		}else{
            Celula tmp = new Celula(x);
            tmp.prox = i.prox;
            i.prox = tmp;
            tmp = i = null;
        }
    }

	public Show removerInicio() throws Exception{
      Show resp;

		if(primeiro == ultimo){
			throw new Exception("Erro ao remover (vazia)!");
		}else{
            primeiro = primeiro.prox;
            resp = primeiro.elemento;
        }

		return resp;
	}

	public Show removerFim() throws Exception{
        Show resp;
        Celula i = null;

		if(primeiro == ultimo){
			throw new Exception("Erro ao remover (vazia)!");
		} else{
            resp = ultimo.elemento;
            for(i = primeiro; i.prox != ultimo; i = i.prox);
            ultimo = i;
            i = ultimo.prox = null;
        }
		return resp;
	}

	public Show removerMeio(int posicao) throws Exception{
        Celula i;
        int cont;
        Show resp;

		if(primeiro == ultimo){
			throw new Exception("Erro ao remover (vazia)!");
        }else{
            for(i = primeiro, cont = 0; (i.prox != ultimo && cont < posicao); i = i.prox, cont++);

		    if(posicao < 0 || posicao > cont + 1){
                throw new Exception("Erro ao remover (posicao " + posicao + " invalida)!");
            } else if(posicao == cont + 1){
                resp = removerFim();
            }else{
                Celula tmp = i.prox;
                resp = tmp.elemento;
                i.prox = tmp.prox;
                tmp.prox = null;
                i = tmp = null;
            }
        }

		return resp;
	}

    public Show remover(Show elemento) throws Exception{
        Show resp;
        Celula ant = primeiro;
        Celula i = primeiro.prox;

        while(i != null && i.elemento != elemento){
            ant = i;
            i = i.prox;
        }

        if(i == null){
            throw new Exception("Erro ao remover (elemento não encontrado)!");
        } else{
            resp = i.elemento;
            ant.prox = i.prox;
            if(i == ultimo){
                ultimo = ant;
            }
            i.prox = null;
        }
        return resp;
    }
}

class Hash{
    private Lista[] tabela;
    private int tamanho;
    private int ultimaPosicaoTentada;

    public Hash(){
        this(21);
    }

    public Hash(int tamanho){
        this.tamanho = tamanho;
        this.tabela = new Lista[tamanho];
        for(int i = 0; i < tamanho; i++){
            tabela[i] = new Lista();
        }
    }

    public int h(Show elemento){
        int soma = 0;
        String titulo = elemento.getTitle();
        for(int i = 0; i < titulo.length(); i++){
            soma += titulo.charAt(i);
        }
        return soma % tamanho;
    }

    public void inserir(Show elemento){
        int pos = h(elemento);
        tabela[pos].inserirInicio(elemento);
    }

    public int pesquisar(Show elemento, Log log){
        int pos = h(elemento);
        ultimaPosicaoTentada = pos;
        boolean encontrado = tabela[pos].pesquisar(elemento, log);
        return (encontrado ? pos : -1);
    }

    public int getUltimaPosicaoTentada(){
        return ultimaPosicaoTentada;
    }
}

class Log{
    public int count;
    public float time;
    public String fileName;
     
    Log(String fileName){
        this.count = 0;
        this.fileName = fileName;
    }
 
    public void incrementaCount(){
        count++;
    }
 
    public void registrarLog(){
        try{
            FileWriter writer = new FileWriter(this.fileName);
            writer.write("Matrícula: 821811 \ttempo de execução: " + this.time + " segundos  \tNúmero de comparações: " + this.count);
            writer.close();

        }catch(IOException e){
            System.out.println("Erro: " + e.getMessage());
        }
    }
}

class Data{
    private String mes;
    private int dia;
    private int ano;

    public Data(){
        this.mes = "NaN";
        this.dia = 0;
        this.ano = 0;
    }

    public Data(String mes, int dia, int ano){
        this.mes = mes;
        this.dia = dia;
        this.ano = ano;
    }

    public String getMes(){
        return mes;
    }
    public void setMes(String mes){
        this.mes = mes;
    }

    public int getDia(){
        return dia;
    }
    public void setDia(int dia){
        this.dia = dia;
    }

    public int getAno(){
        return ano;
    }
    public void setAno(int ano){
        this.ano = ano;
    }

    public String toString() {
        return mes + " " + dia + ", " + ano;
    }

    public static Data Converter(String data){
        String mes = "NaN";
        int dia = 0;
        int ano = 0;
    
        if(!data.equals("NaN")){
            String[] partes = data.split(" ");
            mes = partes[0];
            dia = Integer.parseInt(partes[1].replace(",", ""));
            ano = Integer.parseInt(partes[2]);
        }
    
        return new Data(mes, dia, ano);
    }    
}

public class Show{
    private String show_id;
    private String type;
    private String title;
    private String director;
    private String[] cast;
    private String country;
    private Data data;
    private String release_year;
    private String rating;
    private String duration;
    private String[] listed_in;

    public Show(){
        this.show_id = "";
        this.type = "";
        this.title = "";
        this.director = "";
        this.cast = new String[]{"NaN"};
        this.country = "";
        this.data = new Data();
        this.release_year = "";
        this.rating = "";
        this.duration = "";
        this.listed_in = new String[]{"NaN"};
    }

    public String getShow_id(){
        return show_id;
    }
    public void setShow_id(String show_id){
        this.show_id = show_id;
    }

    public String getType(){
        return type;
    }
    public void setType(String type){
        this.type = type;
    }

    public String getTitle(){
        return title;
    }
    public void setTitle(String title){
        this.title = title;
    }

    public String getDirector(){
        return director;
    }
    public void setDirector(String director){
        this.director = director;
    }

    public String[] getCast(){
        return cast;
    }
    public void setCast(String[] cast){
        this.cast = cast;
    }

    public String getCountry(){
        return country;
    }
    public void setCountry(String country){
        this.country = country;
    }

    public Data getData(){
        return data;
    }
    public void setData(Data data){
        this.data = data;
    }

    public String getRelease_year(){
        return release_year;
    }
    public void setRelease_year(String release_year){
        this.release_year = release_year;
    }

    public String getRating(){
        return rating;
    }
    public void setRating(String rating){
        this.rating = rating;
    }

    public String getDuration(){
        return duration;
    }
    public void setDuration(String duration){
        this.duration = duration;
    }

    public String[] getListed_in(){
        return listed_in;
    }
    public void setListed_in(String[] listed_in){
        this.listed_in = listed_in;
    }

    public Show clone(){
    String[] castClone = (this.cast != null) ? this.cast.clone() : null;
    String[] listedInClone = (this.listed_in != null) ? this.listed_in.clone() : null;
    String dataString = (this.data != null) ? this.data.toString() : null;

    return new Show(this.show_id, this.type, this.title, this.director,
                    castClone, this.country, dataString, this.release_year,
                    this.rating, this.duration, listedInClone);
    }

    public Show(String show_id, String type, String title, String director, String[] cast, String country, String date_added,
                String release_year, String rating, String duration, String[] listed_in){
        this.show_id = show_id;
        this.type = type;
        this.title = title;
        this.director = director;
        this.cast = cast;
        this.country = country;
        this.data = Data.Converter(date_added);
        this.release_year = release_year;
        this.rating = rating;
        this.duration = duration;
        this.listed_in = listed_in;
    }

    public static void ordenar(String[] array){
        for(int i = 0; i < array.length - 1; i++){
            for(int j = 0; j < array.length - i - 1; j++){
                if(array[j].compareTo(array[j + 1]) > 0){
                    String temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
    }    

    public static Show Ler(String line){
        Show s = new Show();
        String[] temp = new String[12];
        int index = 0;
        boolean dentroAspas = false;
        StringBuilder campo = new StringBuilder();
    
        for(int i = 0; i < line.length(); i++){
            char c = line.charAt(i);
    
            if(c == '\"'){
                dentroAspas = !dentroAspas;
            } else if(c == ',' && !dentroAspas){
                temp[index++] = campo.length() == 0 ? "NaN" : campo.toString();
                campo.setLength(0); 
            } else{
                campo.append(c);
            }
        }
    
        temp[index] = campo.length() == 0 ? "NaN" : campo.toString();
    
        s.show_id      = temp[0];
        s.type         = temp[1];
        s.title        = temp[2];
        s.director     = temp[3];
        s.cast = temp[4].equals("NaN") ? new String[]{"NaN"} : temp[4].split(",\\s*");
        if(!temp[4].equals("NaN")){
            ordenar(s.cast);
        }
        s.country      = temp[5];
        s.data         = Data.Converter(temp[6]);
        s.release_year = temp[7];
        s.rating       = temp[8];
        s.duration     = temp[9];
        s.listed_in = temp[10].equals("NaN") ? new String[]{"NaN"} : temp[10].split(",\\s*");
        if(!temp[10].equals("NaN")){
            ordenar(s.listed_in);
        }
    
        return s;
    }
     
    public void Imprimir(){
        System.out.print("=> " + show_id + " ## ");
        System.out.print(title + " ## ");
        System.out.print(type + " ## ");
        System.out.print(director + " ## ");
    
        System.out.print("[");
        if(cast.length == 1 && cast[0].equals("NaN")){
            System.out.print("NaN");
        } else {
            for(int i = 0; i < cast.length; i++){
                System.out.print(cast[i]);
                if(i < cast.length - 1) System.out.print(", ");
            }
        }
        System.out.print("] ## ");
    
        System.out.print(country + " ## ");
        System.out.print(data.toString() + " ## ");
        System.out.print(release_year + " ## ");
        System.out.print(rating + " ## ");
        System.out.print(duration + " ## ");
    
        System.out.print("[");
        if(listed_in.length == 1 && listed_in[0].equals("NaN")){
            System.out.print("NaN");
        } else {
            for(int i = 0; i < listed_in.length; i++){
                System.out.print(listed_in[i]);
                if(i < listed_in.length - 1) System.out.print(", ");
            }
        }
        System.out.println("] ## ");
    }       

    public static void Carregar(Show[] shows){
        File file = new File("/tmp/disneyplus.csv");

        try{
            Scanner sc = new Scanner(file);
            sc.nextLine();
    
            for(int i = 0; i < 1368; i++){
                String linha = sc.nextLine();
                shows[i] = Show.Ler(linha);
            }
    
            sc.close();
        } catch(FileNotFoundException e){
            System.out.println("Arquivo não encontrado!");
        }
    }

    public static int IsFim(String line){
        int result = 1; 

        if(line.charAt(0) == 'F' && line.charAt(1) == 'I' && line.charAt(2) == 'M'){
            result = 0;
        }

        return result;
    }

    public static void swap(Show s[], int i, int j){
        Show tmp = s[i];
        s[i] = s[j];
        s[j] = tmp;
    }

    public static Show Procurar(Show p[], String id){
        Show x = null;

        for(int i = 0; i < p.length && x == null; i++){
            if(p[i] != null && p[i].getShow_id().equals(id)){
                x = p[i];
            }
        }

        return x;
    }

    public static void inserirHash(Hash hash, Show p[], Scanner sc) throws Exception{
        String id = sc.nextLine();

        while(!id.equals("FIM")){
            Show novo;
            novo = Procurar(p, id);
            hash.inserir(novo);
            id = sc.nextLine();
        }
    }

    public static void procuraNaHash(Hash hash, Show[] p, Log log, Scanner sc){
        String titulo = sc.nextLine();

        while (!titulo.equals("FIM")){ 
            Show temp = new Show();
            temp.setTitle(titulo);

            int pos = hash.pesquisar(temp, log);
            System.out.println(" (Posicao: " + (pos == -1 ? hash.getUltimaPosicaoTentada() : pos) + ") " + (pos == -1 ? "NAO" : "SIM"));
            titulo = sc.nextLine(); 
        }
    }

    public static void main(String[] args) throws Exception{
        Scanner sc = new Scanner(System.in);
        Show[] shows = new Show[1368];
        Hash hash = new Hash();
        Log log = new Log("/tmp/821811_hashIndireta.txt");

        Carregar(shows);
        inserirHash(hash, shows, sc);
        Instant start = Instant.now();
        procuraNaHash(hash, shows, log, sc);
        Instant end = Instant.now();
        long elapsedTime = Duration.between(start, end).toMillis();
        log.time = (float) elapsedTime / 1000; 
        log.registrarLog();
        sc.close();
    }
}